bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,212 @@
+            let current_hash = self
+                .inner
+                .block_header(existing_height as usize)?
+                .block_hash();
+            sparse_chain
+                .insert_checkpoint(BlockId {
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("This never errors because we are working with a fresh chain");
+
+            if current_hash == existing_hash {
+                break;
+            }
+        }
+
+        // Insert the new tip so new transactions will be accepted into the sparse chain.
+        let (tip_height, tip_hash) = self.get_tip()?;
+        if let Err(e) = sparse_chain.insert_checkpoint(BlockId {
+            height: tip_height,
+            hash: tip_hash,
+        }) {
+            match e {
+                sparse_chain::InsertCheckpointErr::HashNotMatching => {
+                    // There has been a re-org before we even begin scanning addresses.
+                    // Just recursively call (this should never happen).
+                    return self.wallet_txid_scan(scripts, stop_gap, local_chain, batch_size);
+                }
+            }
+        }
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        for (keychain, mut scripts) in scripts {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+
+            loop {
+                let mut next_batch = (0..batch_size)
+                    .map(|_| scripts.next())
+                    .filter_map(|item| item)
+                    .peekable();
+
+                if next_batch.peek().is_none() {
+                    break;
+                }
+
+                let (indexes, scripts): (Vec<_>, Vec<_>) = next_batch.unzip();
+
+                for (history, index) in self
+                    .batch_script_get_history(scripts.iter())?
+                    .into_iter()
+                    .zip(indexes)
+                {
+                    let txid_list = history
+                        .iter()
+                        .map(|history_result| {
+                            if history_result.height > 0
+                                && (history_result.height as u32) <= tip_height
+                            {
+                                (
+                                    history_result.tx_hash,
+                                    TxHeight::Confirmed(history_result.height as u32),
+                                )
+                            } else {
+                                (history_result.tx_hash, TxHeight::Unconfirmed)
+                            }
+                        })
+                        .collect::<Vec<(Txid, TxHeight)>>();
+
+                    if txid_list.is_empty() {
+                        unused_script_count += 1;
+                    } else {
+                        last_active_index = index;
```

Comment:
Given the API, it is not guaranteed that the scripts will be given in order by index (`impl Iterator<Item = (u32, Script)>`). Should we check if `index` is greater than `last_active_index` before changing?